sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -3168,30 +3282,24 @@ pi_result piQueueRelease(pi_queue Queue) {
         return Res;
 
       // Make sure all commands get executed.
-      // Only do so for a healthy queue as otherwise sync may not be valid.
-      if (Queue->Healthy) {
-        for (auto &ZeQueue : Queue->ComputeQueueGroup.ZeQueues) {
-          if (ZeQueue)
-            ZE_CALL(zeHostSynchronize, (ZeQueue));
-        }
-        for (auto &ZeQueue : Queue->CopyQueueGroup.ZeQueues) {
-          if (ZeQueue)
-            ZE_CALL(zeHostSynchronize, (ZeQueue));
-        }
-      }
+      Queue->synchronize();
 
       // Destroy all the fences created associated with this queue.
       for (auto it = Queue->CommandListMap.begin();
            it != Queue->CommandListMap.end(); ++it) {
         // This fence wasn't yet signalled when we polled it for recycling
         // the command-list, so need to release the command-list too.
-        if (it->second.InUse) {
+        // For immediate commandlists we don't need to do an L0 reset of the
+        // commandlist but do need to do event cleanup which is also in the
+        // resetCommandList function.
+        if (UseImmediateCommandLists || it->second.InUse) {
```

Comment:
Why `InUse` is not enough to check? Can we make the immediate command-list follow the existing semantics of `InUse`, i.e. set to true on a command-list that wasn't reset (including events clean up)?

Reply:
The InUse flag is associated with the fence (see comments in the .hpp file). Taking it over for other uses may cause confusion.

Reply:
The goal of `inUse` was to track command-lists that are still not completed. Regular command-lists do use fences, immediate ones do not. Let's follow this semantics and update comment.

Reply:
OK, comment updated.

Reply:
Hm, but this code should be just this then:
```suggestion
        if (it->second.InUse) {
```

Reply:
Currently (because I believed the InUse flag was for the fence) this flag is initialized to false. If we decide that InUse is actually a representation of the commandlist being in use or not, then for immediate commandlist we could define InUse as always true or always false. Its not really a list so is it in use or not? I could go either way.

Reply:
I think it is really a list both from L0 perspective and PI perspective (pi_command_list_info_t), so please use `InUse` == true through out the duration of immediate command-list lifetime.

Reply:
OK, I'll mark InUse as always true for immediate commandlists.